### PR TITLE
Improve handling of immediate values

### PR DESCRIPTION
**Pull request by Aatch**
The primary improvement here is to handle function calls that return
imemdiate values such that it doesn't force the use of a stack slot. It
also avoids loads from slots that have an store to it in the same block.

This especially improves the codegen for intrinsics, as most of them
produce immediate values.

cc @rust-lang/compiler

### CODE REVIEW

File: src/librustc_trans/trans/base.rs [hunk 1]
```diff
@@ -722,13 +736,20 @@ pub fn invoke<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
...
                               landing_pad,
                               Some(attributes),
                               debug_loc);
+        if let ty::FnConverging(ty) = ret_ty {
+            if return_type_is_void(bcx.ccx(), ty) {
+                llresult = C_nil(bcx.ccx());
+            }
+        } else {
+            llresult = C_nil(bcx.ccx());
+        }
```
Comment (dotdash) [inline src/librustc_trans/trans/base.rs]: I'd prefer to add that check after the if-else block, so it can be merged with the same one in the else-arm.

File: src/librustc_trans/trans/base.rs [hunk 2]
```diff
@@ -775,15 +803,33 @@ pub fn load_if_immediate<'blk, 'tcx>(cx: Block<'blk, 'tcx>,
...
 
     let ptr = to_arg_ty_ptr(cx, ptr, t);
+    let ptr = Value(ptr);
     let align = type_of::align_of(cx.ccx(), t);
 
+    let bb = BasicBlock(cx.llbb);
+
     if type_is_immediate(cx.ccx(), t) && type_of::type_of(cx.ccx(), t).is_aggregate() {
-        let load = Load(cx, ptr);
+        if let Some(val) = ptr.get_dominating_store(cx) {
```
Comment (dotdash) [inline src/librustc_trans/trans/base.rs]: I wonder if this whole block should get its own function, like `fn get_stored_value_opt(alloca: Value) -> Option<Value>` or something like that.

Comment (dotdash) [inline src/librustc_trans/trans/base.rs]: Oops, sent too early. Using a function would avoid the code duplication with the same block below, and would allow us to keep using a `ValueRef` in this function, saving all the `.get()` noise.

File: src/librustc_trans/trans/base.rs [hunk 3]
```diff
@@ -848,7 +918,16 @@ pub fn from_arg_ty(bcx: Block, val: ValueRef, ty: Ty) -> ValueRef {
 
 pub fn to_arg_ty(bcx: Block, val: ValueRef, ty: Ty) -> ValueRef {
     if ty.is_bool() {
-        Trunc(bcx, val, Type::i1(bcx.ccx()))
+        let val = Value(val);
+        if let Some(zext) = val.as_zext_inst() {
+            if let Some(val) = zext.get_operand(0) {
+                let valty = val_ty(val.get());
+                if valty == Type::i1(bcx.ccx()) {
+                    return val.get();
```
Comment (dotdash) [inline src/librustc_trans/trans/base.rs]: Is this worth the complexity? A extend/truncate pair seems like something that neither LLVM nor CPUs (for debug builds) should spend any significant amount of time on.

I guess I might have written the same thing, but looking at it, it just seems so complex.

File: src/librustc_trans/trans/builder.rs [hunk 4]
```diff
@@ -499,9 +507,6 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
     }
 
     pub fn volatile_store(&self, val: ValueRef, ptr: ValueRef) -> ValueRef {
-        debug!("Store {} -> {}",
```
Comment (dotdash) [inline src/librustc_trans/trans/builder.rs]: Why is this removed?

File: src/librustc_trans/trans/callee.rs [hunk 5]
```diff
@@ -731,9 +754,7 @@ pub fn trans_call_inner<'a, 'blk, 'tcx, F>(bcx: Block<'blk, 'tcx>,
         }
     });
 
-    let mut llresult = unsafe {
-        llvm::LLVMGetUndef(Type::nil(ccx).ptr_to().to_ref())
-    };
+    let llresult;
```
Comment (dotdash) [inline src/librustc_trans/trans/callee.rs]: This seems unused

File: src/librustc_trans/trans/expr.rs [hunk 6]
```diff
@@ -2031,18 +2092,36 @@ fn trans_imm_cast<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
...
+            if datum.kind.is_by_ref() {
+                return DatumBlock::new(bcx, Datum::new(
+                    PointerCast(bcx, datum.val, ll_t_out.ptr_to()),
+                    t_out,
+                    Rvalue::new(ByRef))).to_expr_datumblock();
+            } else {
+                let ftys = ll_t_out.field_types();
+
+                let val = C_undef(ll_t_out);
+                let val = InsertValue(bcx, val,
```
Comment (dotdash) [inline src/librustc_trans/trans/expr.rs]: I think `InsertValue` isn't handled by FastISel, so I think @pcwalton won't like this.

**Comment by dotdash** [comment]
:+1: I wanted to get rid of the store/load combo for function calls for sooooo long.

I'm not sure about the load elision thing. How often / in which situations does that come into play?

**Comment by brson** [comment]
Neat. Do compile times improve?

File: src/librustc_trans/trans/base.rs (hunk 3 above)
Comment (Aatch) [inline src/librustc_trans/trans/base.rs]: The primary reason for this change is so the likely/unlikely intrinsics actually work. LLVM is very particular about it and requires that the result from the intrinsic go straight into the condition for the branch. The changes to function calls handles straight `if likely(cond)`, but no `if unsafe { likely(cond) }`. This, combined with the load elision, means that first the load from the temporary is skipped, then the truncation of `zext` bool is skipped.

File: src/librustc_trans/trans/builder.rs (hunk 4 above)
Comment (Aatch) [inline src/librustc_trans/trans/builder.rs]: Because I thought I had added to start with, for some reason...

File: src/librustc_trans/trans/expr.rs (hunk 6 above)
Comment (Aatch) [inline src/librustc_trans/trans/expr.rs]: This shouldn't come up too often though, as expressions like `a as *const [u8]` go through the by-ref path, not the by-value path.

**Comment by Aatch** [comment]
Updated to address comments.

@brson I haven't tested, but my previous attempt at this suggested, no. At least, nothing that isn't covered by noise. The main reason presumably being that `let a = foo()` and similar expressions still compile to about the same code, which probably accounts for a fair amount of function calls.

**Comment by jroesch** [comment]
@Aatch read the patch over and it looks good to me, not sure if you want someone who spends more time in trans to look it over, otherwise r=me.

**Comment by Manishearth** [comment]
*might* be causing http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/6075

**Comment by jroesch** [comment]
@Aatch ping, whats the status on this?

**Comment by dotdash** [comment]
@Aatch Are you ok with me finished this in a new PR?

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with a rebase! (also looks like @jroesch or @dotdash may be willing to help out on this)
